python/cudf/cudf/core/buffer.py

```diff
@@ -1,218 +1,61 @@
-
-        Parameters
-        ----------
-        serialize : callable
-             Used to serialize data that needs serialization .
-        context : dict; optional
-            If not ``None``, it contains information about the destination.
-
-        Returns
-        -------
-        (header, frames)
-            See custom serialization documentation in dask.distributed.
-        """
-        header = {}
-        header["type"] = pickle.dumps(type(self))
-        header["cuda_array_interface"] = self.mem.__cuda_array_interface__
-        return header, [self.mem]
-
-    @classmethod
-    def deserialize(cls, header, frames):
-        """Called when dask.distributed is performing a deserialization for
-        data of this class.
-
-        Do not use this directly.  It is invoked by dask.distributed.
-
-        Parameters
-        ----------
-
-        deserialize : callable
-             Used to deserialize data that needs further deserialization .
-        header, frames : dict
-            See custom serialization documentation in dask.distributed.
-
-        Returns
-        -------
-        obj : Buffer
-            Returns an instance of Buffer.
-        """
-        assert len(frames) == 1, "Use Buffer.serialize() for serialization"
-        iface = header["cuda_array_interface"]
-        if (
-            len(iface["shape"]) == 0 or 0 in iface["shape"]
-        ):  # The array is empty
-            arr = rmm.device_array(
-                iface["shape"], dtype=np.dtype(iface["typestr"])
-            )
+class Buffer:
+    def __init__(self, ptr, size=None, owner=None):
+        if not ptr:
+            self.ptr = 0
+            self.size = 0
         else:
-            # Updating the data pointer to the frame data.
-            iface["data"] = frames[0].__cuda_array_interface__["data"]
-
-            # We need an object that exposes __cuda_array_interface__
-            _dummy_iface = types.SimpleNamespace()
-            _dummy_iface.__cuda_array_interface__ = iface
-
-            # Allocating a new RMM CUDA array with shape and dtype as specified
-            # in `iface` and copy the frame data to it, which makes the memory
-            # survive until the new Buffer goes out of scope.
-            # TODO: when DASK supports RMM, we can simply return it as is.
-            arr, _ = rmm.auto_device(_dummy_iface)
-        return Buffer(arr)
+            self.ptr = ptr
+            self.size = size
+        self._owner = owner
 
     def __reduce__(self):
-        cpumem = self.to_array()
-        # Note: pickled Buffer only stores *size* element.
-        return type(self), (cpumem,)
+        return self.__class__.from_array_like, (self.to_host_array(),)
 
-    def __sizeof__(self):
-        return int(self.mem.alloc_size)
+    def to_host_array(self):
+        return rmm.device_array_from_ptr(
+            self.ptr, nelem=self.size, dtype="int8"
+        ).copy_to_host()
```

Comment:
Would it make more sense to return a Python `bytes` object or a `memoryview` here? I kinda dislike falsely typing it as `int8` here but it would make it harder to implement.

Reply:
We can either use `try(memoryview(obj..)) except`, or at the Cython level use https://docs.python.org/3.6/c-api/buffer.html#c.PyObject_CheckBuffer

Reply:
What if `rmm::device_buffer` handled this for us and we exposed it as a method in `DeviceBuffer`?

xref: https://github.com/rapidsai/rmm/issues/216

Reply:
We don't necessarily have a `device_buffer` underneath a `Buffer` so we can't rely on that. It needs to be done in an agnostic way that any object supporting `__cuda_array_interface__` can support.

Reply:
We could expose cudaMemCopy via RMM Cython if we wanted to be able to do this.

Reply:
This can be done after https://github.com/rapidsai/rmm/pull/219